src/translator/response_builder.cpp

```diff
@@ -9,6 +9,77 @@ void ResponseBuilder::buildQualityScores(Histories &histories, Response &respons
+  std::vector<size_t> tokenOffsetTable(response.source.numSentences(), 0);
+  size_t numSourceTokens = response.source.numWords(0);
+  for (size_t i = 1; i < response.source.numSentences(); i++) {
+    tokenOffsetTable[i] = tokenOffsetTable[i - 1] + response.source.numWords(i);
+    numSourceTokens += response.source.numWords(i);
+  }
+  ABORT_IF(numSourceTokens == 0, "No token in source");
+
+  std::vector<size_t> targetTokenSidTable;
+  std::vector<size_t> targetTokenTidTable;
+  size_t numTargetTokens = 0;
+  for (size_t i = 0; i < response.target.numSentences(); i++) {
+    for (size_t j = 0; j < response.target.numWords(i); j++) {
+      targetTokenSidTable.push_back(i);
+      targetTokenTidTable.push_back(j);
+    }
+    numTargetTokens += response.target.numWords(i);
+  }
+
+  std::vector<size_t> char2TokenTable(response.source.text.size(), 0);
+  std::vector<bool> char2TokenTableValid(response.source.text.size(), 0);
+  for (size_t sentenceId = 0; sentenceId < response.source.numSentences(); sentenceId++) {
+    // Step 0: create or update char-> token table
+    for (size_t s = 0; s < response.source.numWords(sentenceId); s++) {
+      auto sourceByteRange = response.source.wordAsByteRange(sentenceId, s);
+      for (size_t cur = sourceByteRange.begin; cur < sourceByteRange.end; cur++) {
+        char2TokenTable[cur] = s + tokenOffsetTable[sentenceId];
+        char2TokenTableValid[cur] = true;
+      }
+    }
+  }
+
+  for (size_t i = 0; i < char2TokenTable.size(); i++) {
+    if (i > 0 && !char2TokenTableValid[i]) char2TokenTable[i] = char2TokenTable[i - 1] + 1;
+  }
+
+  // Step 1: convert char indices to token indices
+  std::vector<ByteRange> tagPosSourceCharLevel = response.source.tagPosition_;
+  std::vector<TokenIndexRange> tagPosSourceTokenLevel;
+  for (size_t tagIdx = 0; tagIdx < tagPosSourceCharLevel.size(); tagIdx++) {
+    size_t charIdxBegin = tagPosSourceCharLevel[tagIdx].begin;
+    size_t charIdxEnd = tagPosSourceCharLevel[tagIdx].end;
+    tagPosSourceTokenLevel.push_back(TokenIndexRange{char2TokenTable[charIdxBegin], char2TokenTable[charIdxEnd]});
+  }
+
+  // Step 2: scale the positions
+  double ratio = (double)numTargetTokens / numSourceTokens;
+  std::vector<TokenIndexRange> tagPosTargetTokenLevel;
+  for (size_t i = 0; i < tagPosSourceTokenLevel.size(); i++) {
+    size_t tokenBegin = (size_t)(tagPosSourceTokenLevel[i].begin * ratio);
+    size_t tokenEnd = (size_t)(tagPosSourceTokenLevel[i].end * ratio);
+    tagPosTargetTokenLevel.push_back(ByteRange{tokenBegin, tokenEnd});
+  }
+
+  // Step 3: convert token-level tags to the character level one
+  std::vector<ByteRange> tagPosTargetCharLevel;
+  for (TokenIndexRange tokenBound : tagPosTargetTokenLevel) {
+    size_t charBeginSid = targetTokenSidTable[tokenBound.begin];
+    size_t charBeginTid = targetTokenTidTable[tokenBound.begin];
+    size_t charEndSid = targetTokenSidTable[tokenBound.end];
```

Comment:
If `tokenBound` is `[begin, end)`, attempting to access `tokenBound.end` when it might be at end of sentnece is potentially a segfault, right?

Reply:
```
Hello World
Source tag - traversal { [0, 12), [4, 12), [4, 12), [10, 12), [12, 12)}

Thread 32 "bergamot-test" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffe7c18700 (LWP 29475)]
0x000055555565286c in marian::bergamot::buildTagAlignment (response=...) at /home/jphilip/code/bergamot-translator/src/translator/response_builder.cpp:74
74          size_t charEndSid = targetTokenSidTable[tokenBound.end];
(gdb) backtrace
#0  0x000055555565286c in marian::bergamot::buildTagAlignment (response=...) at /home/jphilip/code/bergamot-translator/src/translator/response_builder.cpp:74
```

Please let me know if I'm configuring my ByteRanges for input wrong. `Hello World\n` is 12 characters.

Reply:
My changes are [here](https://github.com/browsermt/bergamot-translator/compare/fake-tag-processor...jerinphilip:fake-tag-processor-delta?expand=1). I have resolved many of the comments above in the mentioned branch already. I'm not sure if I'm inputting wrong, but even if I am, it shouldn't segfault - it should ABORT saying ByteRanges incompatible or something.